### DISCUSSION TITLE
Unit testing?

### DISCUSSION THREAD

**Discussion by tower120** [bevyengine/bevy#2481] [General]
If bevy could support some form of unit-testing out of the box it would be nice.

For example, I have simulation as separate plugin/crate. And I imagine test process as : in that crate "test" make bevy::app, spawn several entities, attach systems, run n-cycles, and then verify that entities have desired values. And I mean native rust-tests.

**Comment by bjorn3** [comment]
I think you can build an app like normally, but only add the necessary plugins and not all default plugins like rendering and windowing. Then you can use `.set_runner` to define a custom runner that runs for the necessary amount of cycles and then returns and finally `.run()` it.

**Reply by tower120** [reply]
Maybe it worth mention that somewhere in the docs.... As a best practices for unit-testing...

_I personally would like to use direct `update` calls instead of fiddling with runner._

**Comment by TheRawMeatball** [comment]
https://github.com/bevyengine/bevy/blob/main/tests/how_to_test_systems.rs

**Reply by tower120** [reply]
Hmm.... Thanks!
